src/Featurizers/TruncatedSVDFeaturizer.h

```diff
@@ -0,0 +1,631 @@
+    T const                                          SingularVectors;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    SVDComponentsAnnotationData(T singularvalues, T singularvectors);
+    ~SVDComponentsAnnotationData(void) = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(SVDComponentsAnnotationData);
+};
+
+namespace ComponentsDetails {
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         SVDTrainingOnlyPolicy
+///  \brief         `SVDComponentsEstimator` implementation details.
+///
+template <typename InputT, typename TransformedT>
+class SVDTrainingOnlyPolicy {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using InputType                                  = InputT;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Data
+    // |
+    // ----------------------------------------------------------------------
+    static constexpr char const * const    NameValue = SVDComponentsEstimatorName;
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    void fit(InputType const &input);
+    SVDComponentsAnnotationData<TransformedT> complete_training(void);
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Data
+    // |
+    // ----------------------------------------------------------------------
+    TransformedT                                     _matrix;
+    TransformedT                                     _singularvalues;
+    TransformedT                                     _singularvectors;
+};
+
+} // namespace ComponentsDetails
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         SVDComponentsEstimator
+///  \brief         This class generates SVDComponentsAnnotationData
+///                 by processing input matrix using Eigen
+///
+template <
+    typename InputT,
+    typename TransformedT,
+    size_t MaxNumTrainingItemsV=std::numeric_limits<size_t>::max()
+>
+class SVDComponentsEstimator :
+    public Components::TrainingOnlyEstimatorImpl<
+        ComponentsDetails::SVDTrainingOnlyPolicy<InputT, TransformedT>,
+        MaxNumTrainingItemsV
+    > {
+public:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Types
+    // |
+    // ----------------------------------------------------------------------
+    using BaseType =
```

Comment:
If there are restrictions on the types, assert that here. If it is possible to derive the TransformedT, do that in the code rather than requesting that information from the user.

Based on our in-person conversation, it sounds like the input is restricted to numeric and integer types. Think also about calculating the TransformedT based on the InputT, thereby eliminating the need for that type.